### DISCUSSION TITLE
Release 3.0.0

### DISCUSSION THREAD

**Discussion by armandino** [instancio/instancio#687] [Announcements]
### New Features

- New `assign()` API that allows setting a value based on generated value. Assignments can be created using static methods provided by `Assign` class:

```java
// Assign Person.name based on the gender field
Person person = Instancio.of(Person.class)
    .assign(given(field(Person::getGender), field(Person::getName))
    .set(When.is(Gender.FEMALE), "Fiona")
    .set(When.is(Gender.MALE), "Michael")
    .elseSet("Oliver"))
    .create();

// Assign Phone.countryCode based on Address.country value
Person person = Instancio.of(Person.class)
    .generate(field(Address::getCountry), gen -> gen.oneOf("CA", "UK", "US"))
    .assign(given(Address::getCountry).isIn("CA", "US").set(field(Phone::getCountryCode), "+1"))
    .assign(given(Address::getCountry).is("UK").set(field(Phone::getCountryCode), "+44"))
    .create();
```

- Support for back-references.

When handling cyclic relationships, the default behaviour of Instancio is to terminate the cycle with a `null` reference. This setting allows setting a back-reference instead of generating `null`:

```java
class Order { List<OrderItem> items; }
class OrderItem { Order order; }

Order order = Instancio.of(Order.class)
    .withSettings(Settings.create().set(Keys.SET_BACK_REFERENCES, true))
    .create();

assertThat(order.getItems()).allSatisfy(item ->
    assertThat(item.getOrder()).isSameAs(order));
```

- Convenience method for creating a `Scope` from a method reference

```java
Select.scope(Person::getWorkAddress)
```

The new API provides an alternative to existing methods:

```java

Select.field(Person::getAddress).toScope();
Select.field(Person.class, "address");
```

- Support for converting a selector `atDepth()` to `Scope`:

```java
all(Foo.class).atDepth(3).toScope();
field(Foo::getBar).atDepth(3).toScope();
```

- Support for scoping `atDepth()` selector using `within()`:

```java
all(Foo.class).atDepth(2).within(scope(Bar.class));
```

- Support for adding selectors `atDepth()` to a group:

```java
TargetSelector group = Select.all(
    all(Foo.class).atDepth(1),
    field(Foo::getBar).atDepth(2),
    all(Baz.class).atDepth(3).within(scope(Jazz.class))
);
```

- `verbose()` method for troubleshooting (outputs current `Settings` and the node hierarchy to standard output):

```java
Person person = Instancio.of(Person.class)
    .verbose()
    .create();
```

- Support for adding predicate selectors to a selector group

```java
Select.all(
    fields(f -> f.getName().equals("foo")),
    fields().annotated(Id.class),
    types().of(Collection.class)
);
```

- Support for `OptionalInt`, `OptionalLong`, and `OptionalDouble` (thanks to @varunu28 for the contribution)
 

### Breaking changes

- Support for metamodels dropped.
  `instancio-processor` was removed, along with `@InstancioMetamodel` annotation.
- Deprecated `values()` attribute was removed from `@InstancioSource` annotation.
The attribute is no longer required since argument types can be resolved from method parameters:

```java
// Before
@InstancioSource({String.class, String.class})
@ParameterizedTest
void example(String first, String second) {}

// After
@InstancioSource
@ParameterizedTest
void example(String first, String second) {}
```

- Deprecated `AsStringGeneratorSpec` interface was removed. The replacement is `AsGeneratorSpec` with a more general-purpose interface.
- Deprecated `GeneratorProvider` SPI was removed. The replacement is `InstancioServiceProvider.GeneratorProvider`.
- Deprecated `TypeResolver` SPI was removed. The replacement is `InstancioServiceProvider.TypeResolver`.
- Settings related to `Keys.ASSIGNMENT_TYPE` were moved from `org.instancio.assignment` to `org.instancio.setting` package.
- Enum `Mode` was moved from `org.instancio` to `org.instancio.settings` package.
- Handling of cyclic classes has been modified. This change affects how cyclic objects are generated
and may result in different object values being produced.

---------------
* Support for assignment and conditional generation (take 3) by @armandino in https://github.com/instancio/instancio/pull/634
* Removed deprecated `instancio-processor` module (#563, #589) by @armandino in https://github.com/instancio/instancio/pull/643
* Removed deprecated `value()` attribute from `InstancioSource` annotation by @armandino in https://github.com/instancio/instancio/pull/644
* Removed deprecated `AsStringGeneratorSpec` by @armandino in https://github.com/instancio/instancio/pull/645
* Removed deprecated `org.instancio.spi.GeneratorProvider` SPI by @armandino in https://github.com/instancio/instancio/pull/646
* Removed deprecated `org.instancio.spi.TypeResolver` SPI by @armandino in https://github.com/instancio/instancio/pull/647
* Updated package name of settings related to `AssignmentType` by @armandino in https://github.com/instancio/instancio/pull/648
* Moved `Mode` enum from `org.instancio` moved to `org.instancio.settings` by @armandino in https://github.com/instancio/instancio/pull/649
* Refactoring, clean up, performance tweaks by @armandino in https://github.com/instancio/instancio/pull/650
* Bug fix: assignment generator not passing on subtype information by @armandino in https://github.com/instancio/instancio/pull/651
* Added `verbose()` method for troubleshooting by @armandino in https://github.com/instancio/instancio/pull/652
* Minor error handling improvements by @armandino in https://github.com/instancio/instancio/pull/653
* Refactor/cleanup node handlers by @armandino in https://github.com/instancio/instancio/pull/654
* Support for scopes when using `atDepth()` with regular selectors by @armandino in https://github.com/instancio/instancio/pull/657
* Convenience method for creating a Scope from a method reference by @armandino in https://github.com/instancio/instancio/pull/658
* Refactor ignored node to retain its type/field by @armandino in https://github.com/instancio/instancio/pull/659
* Removed a conditional branch that never executed. by @armandino in https://github.com/instancio/instancio/pull/660
* Bump equalsverifier-nodep from 3.14.2 to 3.14.3 by @dependabot in https://github.com/instancio/instancio/pull/661
* Bump checkstyle from 10.12.0 to 10.12.1 by @dependabot in https://github.com/instancio/instancio/pull/662
* Support for setting back-references to the `root()` by @armandino in https://github.com/instancio/instancio/pull/664
* Propagate error thrown by custom generator or supplier by @armandino in https://github.com/instancio/instancio/pull/666
* Fix test faliing on Windows due to carriage return by @armandino in https://github.com/instancio/instancio/pull/667
* 519: Generator for OptionalInt by @varunu28 in https://github.com/instancio/instancio/pull/663
* 517: Generator for OptionalLong by @varunu28 in https://github.com/instancio/instancio/pull/668
* Improved error handling of invalid subtype provided by a generator by @armandino in https://github.com/instancio/instancio/pull/670
* 518: Generator for OptionalDouble by @varunu28 in https://github.com/instancio/instancio/pull/672
* Support for customising values of `Optional{Int,Long,Double}` by @armandino in https://github.com/instancio/instancio/pull/674
* Support for adding predicate selectors to a selector group by @armandino in https://github.com/instancio/instancio/pull/671
* Error-handling improvements related to `oneOf()` API. by @armandino in https://github.com/instancio/instancio/pull/675
* Misc refactoring of type utils + tests by @armandino in https://github.com/instancio/instancio/pull/676
* Removed unused code by @armandino in https://github.com/instancio/instancio/pull/677
* Reflection utils cleanup + tests by @armandino in https://github.com/instancio/instancio/pull/678
* Updated OSGi bundle config by @armandino in https://github.com/instancio/instancio/pull/679
* Improvements to error messages from `onComplete()` callbacks by @armandino in https://github.com/instancio/instancio/pull/681
* Bump version.kotlin from 1.8.22 to 1.9.0 by @dependabot in https://github.com/instancio/instancio/pull/680
* Bump guava from 31.1-jre to 32.1.1-jre by @dependabot in https://github.com/instancio/instancio/pull/673
* Collection/map population improvements by @armandino in https://github.com/instancio/instancio/pull/682
* Support for back references using `SET_BACK_REFERENCES` setting by @armandino in https://github.com/instancio/instancio/pull/684
* Better error-handling related to arrays, collections, and maps by @armandino in https://github.com/instancio/instancio/pull/685
* Bump equalsverifier-nodep from 3.14.3 to 3.15 by @dependabot in https://github.com/instancio/instancio/pull/683
* Type mismatch error message improvements by @armandino in https://github.com/instancio/instancio/pull/686

## New Contributors
* @varunu28 made their first contribution in https://github.com/instancio/instancio/pull/663

**Full Changelog**: https://github.com/instancio/instancio/compare/instancio-parent-2.16.1...instancio-parent-3.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/instancio/instancio/releases/tag/instancio-parent-3.0.0'>3.0.0</a>.</em>

**Comment by nickngn** [comment]
Awesome, assign() API is what I'm looking forward to...
